File: is-graph-bipartite/is-graph-bipartite.cpp
class Solution {
    bool check(int n,int start,vector<vector<int>> &graph,vector<int> &color)
    {
        color[start] = 1;
        queue<int> q;
        q.push(start);
        while(!q.empty())
        {
            int node = q.front();
            q.pop();
            for(int ele : graph[node])
            {
                if(color[ele] == -1)
                {
                    color[ele] = 1 - color[node];
                    q.push(ele);
                }
                else if(color[ele] == color[node]) return false;
            }
        }
        return true;
    }
    
public:
    bool isBipartite(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<int> color(n,-1);
        for(int i=0;i<n;i++)
        {
            if(color[i] == -1)
            {
                if(!check(n,i,graph,color)) return false;
            }
        }
        return true;
    }
};

File: is-graph-bipartite/is-graph-bipartite.java
class Solution {
    public boolean isBipartite(int[][] graph) {
        int n = graph.length;
        int []color = new int[n];
        Arrays.fill(color,-1);

        
        Queue<Integer>qu = new LinkedList<>();
        for(int i=0;i<n;i++)
        {
            if(color[i] != -1) continue;
            qu.add(i);
            color[i]=0;

            while(!qu.isEmpty()){
                int top = qu.poll();
                int col = color[top];

                for(int x: graph[top]){
                    if(color[x]==-1)
                        {qu.add(x); color[x]=1-col;}
                    else if(color[x]==color[top]) return false;
                }
            }
        }
        
        return true;
    }
}

